pageserver/src/tenant/mgr.rs

```diff
@@ -864,34 +876,185 @@ pub(crate) enum GetTenantError {
+async fn timeout_cancellable<F>(
+    duration: Duration,
+    future: F,
+    cancel: &CancellationToken,
+) -> Result<F::Output, TimeoutCancellableError>
+where
+    F: std::future::Future,
+{
+    tokio::select!(
+        r = tokio::time::timeout(duration, future) => {
+            r.map_err(|_| TimeoutCancellableError::Timeout)
+
+        },
+        _ = cancel.cancelled() => {
+            Err(TimeoutCancellableError::Cancelled)
+
+        }
+    )
+}
+
+/// Get a [`Tenant`] in its active state. If the tenant_id is currently in [`TenantSlot::InProgress`]
+/// state, then wait for up to `timeout`.  If the [`Tenant`] is not currently in [`TenantState::Active`],
+/// then wait for up to `timeout` (minus however long we waited for the slot).
+pub(crate) async fn get_active_tenant_with_timeout(
+    tenant_id: TenantId,
+    mut timeout: Duration,
+    cancel: &CancellationToken,
+) -> Result<Arc<Tenant>, GetActiveTenantError> {
+    enum WaitFor {
+        Barrier(utils::completion::Barrier),
+        Tenant(Arc<Tenant>),
+    }
+
+    let wait_for = {
+        let locked = TENANTS.read().unwrap();
+        let peek_slot =
+            tenant_map_peek_slot(&locked, &tenant_id, true).map_err(GetTenantError::MapState)?;
+        match peek_slot {
+            Some(TenantSlot::Attached(tenant)) => {
+                match tenant.current_state() {
+                    TenantState::Active => {
+                        // Fast path: we don't need to do any async waiting.
+                        return Ok(tenant.clone());
+                    }
+                    _ => WaitFor::Tenant(tenant.clone()),
+                }
+            }
+            Some(TenantSlot::Secondary) => {
+                return Err(GetActiveTenantError::NotFound(GetTenantError::NotActive(
+                    tenant_id,
+                )))
+            }
+            Some(TenantSlot::InProgress(barrier)) => WaitFor::Barrier(barrier.clone()),
+            None => {
+                return Err(GetActiveTenantError::NotFound(GetTenantError::NotFound(
+                    tenant_id,
+                )))
+            }
+        }
+    };
+
+    let tenant = match wait_for {
+        WaitFor::Barrier(barrier) => {
+            tracing::debug!("Waiting for tenant InProgress state to pass...");
+            let wait_start = Instant::now();
+            timeout_cancellable(timeout, barrier.wait(), cancel)
+                .await
+                .map_err(|e| match e {
+                    TimeoutCancellableError::Timeout => {
+                        GetActiveTenantError::WaitForActiveTimeout {
+                            latest_state: None,
+                            wait_time: wait_start.elapsed(),
+                        }
+                    }
+                    TimeoutCancellableError::Cancelled => GetActiveTenantError::Cancelled,
+                })?;
+            let wait_duration = Instant::now().duration_since(wait_start);
+            timeout -= wait_duration;
```

Comment:
The better solution here is to compute a deadline at the top of the function, and use `until()` for to compute the timeout arguments to sub-functions. (Really, we should be using deadlines more)

Reply:
This is true.  Refactored to use a deadline.  8d6b95b44